agent/fw_drupal8.c

```diff
@@ -406,9 +440,24 @@ NR_PHP_WRAPPER(nr_drupal94_invoke_all_with_callback) {
     goto leave;
   }
 
+#if ZEND_MODULE_API_NO >= ZEND_8_0_X_API_NO \
+    && !defined OVERWRITE_ZEND_EXECUTE_DATA
+  zval* curr_hook
+      = (zval*)nr_stack_get_top(&NRPRG(drupal_invoke_all_hooks));
```

Comment:
Should the `states` stack be queried first to see if `hook` name can be popped?

Reply:
It is rather confusing, and part of why I've abstracted management of the stacks into a separate API. The states stack keeps track whether of not the current invoke_all* execution successfully pushed a hook onto the hooks stack. If it failed to add the hook (`nr_php_zval_is_non_empty_string` returned `false`) then the states stack reflects this with a false value. If it succeeded, the state stack reflects it with a true value. In all cases, the states stack will match our depth of invoke_all* calls, while the hooks stack might not.

So no, we do not need to query the states stack to peek at the top of the hooks stack. Because if the top of the states stack were false, that is simply telling us that the top of the hook stack is from an upstream nesting execution of invoke_all*, or its NULL. Either way is fine. This functionality is inline with how it was handled pre-OAPI.

The real purpose of the states stack is to facilitate communication from the _before callback to the _after callback. At that time, we free the top of the hooks stack if and only if the top of the states stack is `true`. This saves us another copy of the hook argument, and another call to `nr_php_zval_is_non_empty_string`.

Reply:
In reality, I am unsure how often nested invoke_all calls really happen

Reply:
On retrospect, the states stack might be needlessly complicated. As stated, the main purpose is to avoid copying the hook argument again. However, the function [nr_php_get_user_func_arg](https://github.com/newrelic/newrelic-php-agent/blob/main/agent/php_agent.c#L417) can accomplish this same goal.

Reply:
Further investigations: during the "_clean" functions, we do not have access to the function arguments and would be unable to call nr_php_get_user_func_arg. Therefore, the state_stack is absolutely needed to properly convey the necessary cleaning information.